### ISSUE TITLE
Schedule Validation

### ISSUE THREAD

**Issue by amyfinck** [SENG-499-Company-3/algorithm-1#19]
Our validation algorithm needs to be updated to account for the time block method we are using. It also needs to check that required courses in the same year are not offered at the same time.

**Comment by c3n0te** [comment]
Action items: 

1. theorize and trace out how to validate course-teacher (min-max teacher course loads), time-teacher (teaching more than one class at the same time) and time-course (more than one required course offered at the same time) hard constraints with less time slots (51 -> 15 core + 5 flex timeslots)
2. implement new validation algorithms with vectorized numpy routines
3. make any necessary api changes (e.g. get information about required-course dependencies)

**Comment by c3n0te** [comment]
### Context 
As a precursor to building our algorithm, it is important to consider the theoretical and practical limitations we expect to face. Notably, our approach, solving the weighted Max-Cut problem on a hypergraph, relies on our ability to represent a schedule as some kind of data structure in which computations can be performed on. Specifically, we choose to represent a schedule as a tensor - a multidimensional array which can be thought of as matrices layered on top of each other. Matrix sandwich anybody????

The hypergraph tensor is defined in R<sup>k</sup>, each dimension in this tensor represents a partitioned (mutually exclusive) subset of entities to be assigned. In the context of scheduling for UVic we have chosen k=3, where the dimensions are courses, times, and teachers; but, it is important to note that this model can support additional dimensions if desired. An assignment in this space is represented as a vector with k elements. Each element of the vector represents a _unique_ entity from the dimension the entity was selected from.

This model is convenient for a variety of reasons, mainly we do not have to spend extra computations partitioning our entities before finding the maximum weights for each partition (i.e. we only perform the weight optimization step, not the cut step), the hypergraph is cut by construction. This is made possible because we have information about our entities prior to creating our hypergraph, i.e. we know a teacher is not the same as a course. If we attempted to solve this problem with a normal graph implemented with an adjacency matrix we would either need to compute the k cuts at run-time or put in some additional thought as to how we could encode the mutual exclusivity of our vertex subsets inside the graph which would also likely add costly jumps between far off memory locations when performing validation (i.e. reduced locality).

Solving this problem with a hypergraph/tensor translates into three practical benefits: 1) elimination of costly cut operations during optimization, 2) increased memory read efficiency due to better cache locality during schedule validation, and 3) increased memory write efficiency by representing a sparse tensor as a Hashtable during object (re-)allocation and projection. 

### tl;dr
    * Computing less is good. 
    * Computing on things which lie close to each other in memory is good. 
    * Representing very large things as smaller things is good.


### Now onto validation...

Cache locality is hard to measure in Python due to memory allocations being abstracted away by the interpreter, but the NumPy package we use to implement our tensor is written in C/C++ which implies that making cache locality a focus should translate into a direct performance boost, especially since schedule validation relies heavily on how quickly we can access and read values stored in memory. See [NumPy performance tips](https://shihchinw.github.io/2019/03/performance-tips-of-numpy-ndarray.html) for more discussion of strategies to increase performance (of particular note are sections titled **Locality Matters** and **Vectorization**).  


Constraint validation will be done as follows: 

1. Course - Teacher: 
Receive information about course loads from API call request body; project 3D tensor (courses, times, teachers) to 2D (teachers, courses), note the order of (x, y) = (teachers, courses) is important to maximize cache efficiency and is row-major, this means each teacher is represented as a row and each course is represented as a column; count the number of courses assigned to each teacher and ensure it is less than or equal to their specified load; count the number of teachers assigned to each course and ensure it is 0 or 1 (course not assigned or course assigned to exactly one teacher, we will not consider the case where one course is taught by multiple professors yet).


2. Time - Teacher: time slots are mutually exclusive (no temporal overlap), this implies that a teacher can never be assigned to two time slots which overlap. However, it is still possible for a teacher to be assigned the same time slot more than once. To satisfy this constraint we count the number of occurrences each teacher has been assigned to each time slot and if it is greater than one the schedule is invalid.


3. Course - Time: receive information about required courses (or any courses that should not be scheduled at the same time) from API call request body in the form of partition indices; ensure courses are ordered according to these partitions (i.e. each partition contains a group of courses which should not be scheduled at the same time and the provided course list is ordered as a sequence of these partitions); project 3D tensor (courses, times, teachers) to 2D (times, courses); define a window as the difference between the current partition index and the next partition index; count the number of courses scheduled to occur in this window and ensure it is 0 or 1; continue sliding the window over each partition and counting courses in the window until you have done this for every partition; then project 3D tensor (courses, times, teachers) to 2D (courses, times); count the number of time slots assigned to each course and ensure it is 0 or 1.


4. A schedule is valid if each of the above hard constraints are satisfied, otherwise the schedule is invalid.

**Comment by c3n0te** [comment]
Some NumPy routines which will be useful include np.nditer (NumPy iterator) to iterate through a NumPy array; np.count_nonzero to count non-zero elements in a NumPy array; and indexing, slicing and broadcasting.
